### PR TITLE
feat: configurable caching for postgres with support for tracing

### PR DESCRIPTION
**Pull request by jacek-prisma**
This PR fixes https://github.com/prisma/team-orm/issues/1331.

### CODE REVIEW

File: quaint/src/pooled/manager.rs [hunk 1]
```diff
@@ -133,9 +134,23 @@ impl Manager for QuaintManager {
...
+
+            #[cfg(feature = "postgresql-native")]
+            QuaintManager::Postgres {
+                url,
+                tls_manager,
+                is_tracing_enabled: true,
+            } => {
+                use crate::connector::PostgreSqlWithTracingCache;
+                Ok(Box::new(PostgreSqlWithTracingCache::new(url.clone(), tls_manager).await?) as Self::Connection)
             }
```
Comment (jkomyno) [inline quaint/src/pooled/manager.rs:154]: Summary:
- when `is_tracing_enabled`, we use the tracing cache (`crate::connector::PostgreSqlWithTracingCache`)
- when `!is_tracing_enabled`, we use the default cache (`crate::connector::PostgreSqlWithDefaultCache`)

File: quaint/src/connector/postgres/native/cache.rs [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+            let stmt3 = cache.get_statement(&client, sql, &types).await.unwrap();
+            assert_ne!(stmt1.name(), stmt3.name());
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn tracing_lru_cache_reuses_queries_within_capacity() {
+        run_with_client(|client| async move {
+            let cache = TracingLruCache::with_capacity(1);
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:310]: Let's expand this test case with a higher capacity, e.g., 3.

File: quaint/src/connector/postgres/native/cache.rs [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+            let stmt3 = cache.get_query(&client, sql, &types).await.unwrap();
+            assert_ne!(stmt1.name(), stmt3.name());
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn prepared_statement_lru_cache_reuses_statements_within_capacity() {
+        run_with_client(|client| async move {
+            let cache = PreparedStatementLruCache::with_capacity(1);
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:289]: Let's expand this test case with a higher capacity, e.g., 3.

File: quaint/src/connector/postgres/native/cache.rs [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+            let stmt2 = cache.get_statement(&client, sql, &types).await.unwrap();
+            assert_ne!(stmt1.name(), stmt2.name());
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn prepared_statement_lru_cache_reuses_queries_within_capacity() {
+        run_with_client(|client| async move {
+            let cache = PreparedStatementLruCache::with_capacity(1);
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:268]: Let's expand this test case with a higher capacity, e.g., 3.

File: quaint/src/connector/postgres/native/cache.rs [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+    #[inline]
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<Statement, Error> {
+        self.get_statement(client, sql, types).await
+    }
+
+    #[inline]
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error> {
+        client.prepare_typed(sql, types).await
+    }
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:49]: Is it possible to get rid of `Result`? It seems to me that, so far, you only use the `Ok()` constructor or `.unwrap()` the result.

```suggestion
    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Statement {
        client.prepare_typed(sql, types).await
    }
```

File: quaint/src/connector/postgres/native/cache.rs [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+pub struct NoOpCache;
+
+#[async_trait]
+impl QueryCache for NoOpCache {
+    type Query<'a> = Statement;
+
+    #[inline]
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<Statement, Error> {
+        self.get_statement(client, sql, types).await
+    }
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:44]: Is it possible to get rid of Result? It seems to me that, so far, you only use the Ok() constructor or .unwrap() the result.

```suggestion
    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Statement {
        self.get_statement(client, sql, types).await
    }
```

File: quaint/src/connector/postgres/native/cache.rs [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error> {
+        match self.cache.get(sql, types).await {
+            Some(statement) => Ok(statement),
+            None => {
+                let stmt = client.prepare_typed(sql, types).await?;
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:85]: Ah, here's why we need `Result` in the function signature's return type

File: quaint/src/connector/postgres/native/cache.rs (hunk 6 above)
Comment (jkomyno) [reply quaint/src/connector/postgres/native/cache.rs:44]: Nevermind, I replied to myself in https://github.com/prisma/prisma-engines/pull/5082/files#r1881880519.

File: quaint/src/connector/postgres/native/cache.rs (hunk 5 above)
Comment (jkomyno) [reply quaint/src/connector/postgres/native/cache.rs:49]: Nevermind, I replied to myself in https://github.com/prisma/prisma-engines/pull/5082/files#r1881880519.

Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs:49]: it's unwrapped in tests, but we do pass the error through in `native/mod.rs`, so unwrapping here would change the behavior

File: quaint/src/connector/postgres/native/cache.rs (hunk 6 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs:44]: same as above

File: quaint/src/connector/postgres/native/cache.rs [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+    /// Retrieve a prepared query.
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<Self::Query<'a>, Error>;
+
+    /// Retrieve a prepared statement.
+    ///
+    /// This is useful in scenarios that require direct access to a prepared statement,
+    /// e.g. describing a query.
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error>;
+}
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/cache.rs:30]: Yes! This is so much better than having two separate caches.

File: quaint/src/connector/postgres/native/cache.rs (hunk 2 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs]: done

File: quaint/src/connector/postgres/native/cache.rs (hunk 3 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs]: done

File: quaint/src/connector/postgres/native/cache.rs (hunk 4 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs]: done

File: quaint/src/connector/postgres/native/cache.rs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+    type Query<'a> = TypedQuery<'a>;
+
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<TypedQuery<'a>, Error> {
+        let sql_without_traceparent = strip_query_traceparent(sql);
+
+        let metadata = match self.cache.get(sql_without_traceparent, types).await {
+            Some(metadata) => metadata,
+            None => {
+                let stmt = client.prepare_typed(sql_without_traceparent, types).await?;
+                let metdata = Arc::new(QueryMetadata::from(&stmt));
```
Comment (aqrln) [inline quaint/src/connector/postgres/native/cache.rs:129]: typo: met**a**data

File: quaint/src/connector/postgres/native/cache.rs [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+    pub capacity: usize,
+}
+
+/// Key uniquely representing an SQL statement in the prepared statements cache.
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct QueryKey {
+    /// Hash of a string with SQL query.
+    sql: u64,
+    /// Combined hash of types for all parameters from the query.
+    types_hash: u64,
```
Comment (aqrln) [inline quaint/src/connector/postgres/native/cache.rs:160]: I don't know much about the underlying math, but is taking a hash of a hash safe in terms of it statistical properties? Should we have a no-op hasher that just combines the inputs for the hash map in the `LruCache` itself?

File: quaint/src/connector/postgres/native/cache.rs [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+        let mut cache = self.cache.lock().await;
+        let capacity = cache.capacity();
+        let stored = cache.len();
+
+        let key = QueryKey::new(&self.state, sql, types);
+        // we call `get_mut` because LRU requires mutable access for lookups
+        match cache.get_mut(&key) {
+            Some(value) => {
+                tracing::trace!(
+                    message = "CACHE HIT!",
```
Comment (aqrln) [inline quaint/src/connector/postgres/native/cache.rs:199]: fwiw these two logs in caps have always annoyed me when I saw them in code 😆

File: quaint/src/connector/postgres/native/cache.rs (hunk 10 above)
Comment (aqrln) [reply quaint/src/connector/postgres/native/cache.rs:160]: (I guess it's probably is and it should be fine, we're just doing slightly more computations than technically necessary)

File: quaint/src/connector/postgres/native/query.rs [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+use postgres_types::{BorrowToSql, Type};
+use tokio_postgres::{Client, Error, RowStream, Statement};
+
+/// Types that can be dispatched to the database as a query and carry the necessary type
+/// information about its parameters and columns to interpret the results.
+#[async_trait]
+pub trait PreparedQuery: Send {
+    fn param_types(&self) -> impl ExactSizeIterator<Item = &Type> + '_;
+    fn column_names(&self) -> impl ExactSizeIterator<Item = &str> + '_;
+    fn column_types(&self) -> impl ExactSizeIterator<Item = &Type> + '_;
```
Comment (aqrln) [inline quaint/src/connector/postgres/native/query.rs:13]: nit (feel free to ignore, and rust 2024 is around the corner anyway so it all will be removed soon)

```suggestion
    fn param_types(&self) -> impl ExactSizeIterator<Item = &Type> + use<'_>;
    fn column_names(&self) -> impl ExactSizeIterator<Item = &str> + use<'_>;
    fn column_types(&self) -> impl ExactSizeIterator<Item = &Type> + use<'_>;
```

File: quaint/src/connector/postgres/native/cache.rs [hunk 13]
```diff
@@ -0,0 +1,394 @@
+use std::{
+    hash::{BuildHasher, Hash, RandomState},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use futures::lock::Mutex;
```
[the remark this answers is not in the document]
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs:7]: I mimicked the existing LRU implementation, so this is just a leftover from that, I don't mind changing the Mutex

File: quaint/src/connector/postgres/native/cache.rs (hunk 11 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs:199]: I can change the message to something less annoying if you have any suggestions 😅

File: quaint/src/connector/postgres/native/query.rs (hunk 12 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/query.rs:13]: I'm getting an error with this syntax, might be some issue with the async trait macro

File: quaint/src/connector/postgres/native/cache.rs (hunk 9 above)
Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs]: fixed

File: quaint/src/connector/postgres/native/query.rs (hunk 12 above)
Comment (aqrln) [reply quaint/src/connector/postgres/native/query.rs:13]: Ah right, precise capturing in traits is still unstable. Actually, do we even need the outlives trick here? RPITIT already uses the new Rust 2024 lifetime capture rules IIRC.

Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/query.rs:13]: you're right, it seems to compile without it, I'll remove it

File: quaint/src/connector/postgres/native/cache.rs (hunk 11 above)
Comment (aqrln) [reply quaint/src/connector/postgres/native/cache.rs:201]: I think just `message = "cache hit"` and `message = "cache miss"` is fine but I don't really care that much, your call 😄

Comment (jacek-prisma) [reply quaint/src/connector/postgres/native/cache.rs]: ok, updated it

**Review by aqrln — APPROVED** [review]
The code looks amazing, and the client tests pass with `pnpm --filter @prisma/client test:functional:code --provider postgresql --preview-features tracing` locally.

The lack of `execute_typed` in `tokio_postgres` is a bummer but I think it only affects `createMany`.

**Pull request merged**